### PR TITLE
Add PySpark support to sparkctl and Spark Operator. 

### PR DESCRIPTION
**Pull request by mrow4a**
## Description
This PR add a support for PySpark, and thus closes https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/181

@liyinan926
CC @prasanthkothuri

### CODE REVIEW

File: examples/spark-pyfiles.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+apiVersion: "sparkoperator.k8s.io/v1alpha1"
+kind: SparkApplication
+metadata:
+  name: pyfiles
+  namespace: default
+spec:
+  type: Python
+  pythonVersion: "2"
+  mode: cluster
+  image: "gcr.io/ynli-k8s/spark:v2.4.0-SNAPSHOT"
```
Comment (mrow4a) [inline examples/spark-pyfiles.yaml]: @liyinan926 I for now placed SNAPSHOT tag of your image.

Comment (liyinan926) [reply examples/spark-pyfiles.yaml]: That's fine. I will make sure the image exists before merging this.

File: docs/user-guide.md [hunk 2]
```diff
@@ -362,6 +369,42 @@ spec:
...
+to run the driver and executor containers. Below is an example showing part of a `SparkApplication` specification:
+
+```yaml
+spec:
+  type: Python
+  pythonVersion: 2
+  mainApplicationFile: local:///opt/spark/examples/src/main/python/pyfiles.py
+```
+
+Some PySpark applications need additional Python packages additionally to the main application resource to run.
```
Comment (liyinan926) [inline docs/user-guide.md]: `additionally` is duplicated and can be removed.

File: docs/user-guide.md [hunk 3]
```diff
@@ -362,6 +369,42 @@ spec:
...
+
+```yaml
+spec:
+  type: Python
+  pythonVersion: 2
+  mainApplicationFile: local:///opt/spark/examples/src/main/python/pyfiles.py
+```
+
+Some PySpark applications need additional Python packages additionally to the main application resource to run. 
+Such dependencies are specified using the `--py-files` option of `spark-submit` command.
```
Comment (liyinan926) [inline docs/user-guide.md]: I would rephrase this sentence as "Such dependencies are specified using the optional field `.spec.deps.pyFiles` , which translates to the `--py-files` option of the `spark-submit` command.".

File: docs/user-guide.md [hunk 4]
```diff
@@ -362,6 +369,42 @@ spec:
...
+    pyfiles:
+       - local:///opt/spark/examples/src/main/python/py_container_checks.py
+       - gs://spark-data/python-dep.zip
+``` 
+
+In order to use the dependencies that are hosted remotely, the following PySpark code
+can be used.
+  
+```
+python_dep_file_path = SparkFiles.get("python-dep.zip")
```
Comment (liyinan926) [inline docs/user-guide.md]: Ditto. Using `SparkFiles.get()` only works for Spark 2.4.

File: docs/api.md [hunk 5]
```diff
@@ -53,6 +54,7 @@ A `SparkApplicationSpec` has the following top-level fields:
 | `NodeSelector` | `spark.kubernetes.node.selector.[labelKey]` | Node selector of the driver pod and executor pods, with key `labelKey` and value as the label's value. |
 | `MaxSubmissionRetries` | N/A | The maximum number of times to retry a failed submission. |
 | `SubmissionRetryInterval` | N/A | The unit of intervals in seconds between submission retries. Depending on the implementation, the actual interval between two submission retries may be a multiple of `SubmissionRetryInterval`, e.g., if linear or exponential backoff is used. |
+| `MemoryOverheadFactor` | `spark.kubernetes.memoryOverheadFactor` | This sets the Memory Overhead Factor that will allocate memory to non-JVM memory. For JVM-based jobs this value will default to 0.10, for non-JVM jobs 0.40. |
```
Comment (liyinan926) [inline docs/api.md]: It's worth mentioning that the value of this field will be overridden by Spec.Driver.MemoryOverhead and Spec.Executor.MemoryOverhead if they are set.

File: pkg/apis/sparkoperator.k8s.io/v1alpha1/types.go [hunk 6]
```diff
@@ -212,6 +212,14 @@ type SparkApplicationSpec struct {
 	// SubmissionRetryInterval is the unit of intervals in seconds between submission retries.
 	// Optional.
 	SubmissionRetryInterval *int64 `json:"submissionRetryInterval,omitempty"`
+	// This sets the major Python version of the docker
+	// image used to run the driver and executor containers. Can either be 2 or 3, default 2.
+	// Optional.
+	PythonVersion *string `json:"pythonVersion,omitempty"`
+	// This sets the Memory Overhead Factor that will allocate memory to non-JVM memory.
```
[the remark this answers is not in the document]
Comment (mrow4a) [reply pkg/apis/sparkoperator.k8s.io/v1alpha1/types.go]: Addressed

**Comment by mrow4a** [comment]
@liyinan926 Addressed required changes.

File: examples/spark-pyfiles.yaml (hunk 1 above)
Comment (mrow4a) [reply examples/spark-pyfiles.yaml]: Great!

File: docs/user-guide.md [hunk 7]
```diff
@@ -362,6 +362,42 @@ spec:
...
+
+In order to use the dependencies that are hosted remotely, the following PySpark code
+can be used in Spark 2.4.
+  
+```
+python_dep_file_path = SparkFiles.get("python-dep.zip")
+spark.sparkContext.addPyFile(dep_file_path)
+``` 
+
+Note that Python binding for PySpark will available in Apache Spark 2.4,
```
Comment (liyinan926) [inline docs/user-guide.md]: `will be available`.

File: docs/user-guide.md [hunk 8]
```diff
@@ -362,6 +362,42 @@ spec:
...
+In order to use the dependencies that are hosted remotely, the following PySpark code
+can be used in Spark 2.4.
+  
+```
+python_dep_file_path = SparkFiles.get("python-dep.zip")
+spark.sparkContext.addPyFile(dep_file_path)
+``` 
+
+Note that Python binding for PySpark will available in Apache Spark 2.4, 
+and currently requires building custom 2.4.0-SNAPSHOT Docker image.
```
Comment (liyinan926) [inline docs/user-guide.md]: `currently requires building a custom Docker image from the Spark master branch.`.

**Comment by mrow4a** [comment]
@liyinan926 done

**Comment by liyinan926** [comment]
Thanks! Will merge once I have the image `gcr.io/ynli-k8s/spark:v2.4.0-SNAPSHOT` pushed.

**Comment by liyinan926** [comment]
@mrow4a Can you change the image to `gcr.io/ynli-k8s/spark-py:v2.4.0-SNAPSHOT`?

**Comment by liyinan926** [comment]
@mrow4a No worries, I will update the image after merging. Thanks!
